storage/cloudspanner/tree_storage.go

```diff
@@ -0,0 +1,523 @@
+// On return from the call, this transaction will be in a closed state.
+func (t *treeTX) Commit() error {
+	t.mu.Lock()
+	defer func() {
+		t.stx = nil
+		t.mu.Unlock()
+	}()
+
+	if t.stx == nil {
+		return ErrTransactionClosed
+	}
+	switch stx := t.stx.(type) {
+	case *spanner.ReadOnlyTransaction:
+		glog.V(1).Infof("Closed readonly tx %p", stx)
+		stx.Close()
+		return nil
+	case *spanner.ReadWriteTransaction:
+		return t.flushSubtrees()
+	default:
+		return fmt.Errorf("internal error: unknown transaction type %T", stx)
+	}
+}
+
+// Rollback aborts any operations perfomed on the underlying Spanner
+// transaction.
+// On return from the call, this transaction will be in a closed state.
+func (t *treeTX) Rollback() error {
+	t.mu.Lock()
+	defer func() {
+		t.stx = nil
+		t.mu.Unlock()
+	}()
+
+	if t.stx == nil {
+		return ErrTransactionClosed
+	}
+	return nil
+}
+
+func (t *treeTX) Close() error {
+	if t.IsOpen() {
+		if err := t.Rollback(); err != nil && err != ErrTransactionClosed {
+			glog.Warningf("Rollback error on Close(): %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+// IsOpen returns true iff neither Commit nor Rollback have been called.
+// If this function returns false, further operations may not be attempted on
+// this transaction object.
+func (t *treeTX) IsOpen() bool {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.stx != nil
+}
+
+// ReadRevision returns the tree revision at which the current STH was stored.
+func (t *treeTX) ReadRevision() int64 {
+	sth, err := t.currentSTH(context.TODO())
+	if err != nil {
+		panic(err)
+	}
+	return sth.TreeRevision
+}
+
+// WriteRevision returns the tree revision at which any tree-modifying
+// operations will write.
+func (t *treeTX) WriteRevision() int64 {
+	rev, err := t.writeRev(context.TODO())
+	if err != nil {
+		panic(err)
+	}
+	return rev
+}
+
+// nodeIDToKey returns a []byte suitable for use as a primary key column for
+// the subtree which contains the id.
+// If id's prefix is not byte-aligned, an error will be returned.
+func subtreeKey(id storage.NodeID) ([]byte, error) {
+	// TODO(al): extend this check to ensure id is at a tree stratum boundary.
+	if id.PrefixLenBits%8 != 0 {
+		return nil, fmt.Errorf("id.PrefixLenBits (%d) is not a multiple of 8; it cannot be a subtree prefix", id.PrefixLenBits)
+	}
+	return id.Path[:id.PrefixLenBits/8], nil
+}
+
+// getSubtree retrieves the subtree specified by id at (or below) the requested revision.
+// If no such subtree exists it returns nil.
+func (t *treeTX) getSubtree(ctx context.Context, rev int64, id storage.NodeID) (p *storagepb.SubtreeProto, e error) {
+	stID, err := subtreeKey(id)
+	if err != nil {
+		return nil, err
+	}
+
+	var ret *storagepb.SubtreeProto
+	prefix := spanner.Key{t.treeID, stID}.AsPrefix()
+	rows := t.stx.Read(ctx, subtreeTbl, prefix, []string{colRevision, colSubtree})
+	err = rows.Do(func(r *spanner.Row) error {
+		var rRev int64
+		var st storagepb.SubtreeProto
+		stBytes := make([]byte, 1<<20)
```

Comment:
Is 1<<20 bigger than needed?

Reply:
I think it'll get reallocated if more space is needed, this should just be big enough for most situations.

Reply:
Yes, it was just that we have this column set to max length 32K in the SQL code so it looked high.

Reply:
Hmm, 32K is probably way too small when maps are involved... you might want to change that!

Reply:
Yes I think it should be increased.

Reply:
Addressing this for MySQL in #1019.